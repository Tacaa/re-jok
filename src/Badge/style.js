import styled, { keyframes} from 'styled-components'

const bouncing = keyframes `
0%{
  transform:scale(.9) translateX(-50%);

}
25%{
  transform:scale(1.1) translateX(-50%);
}
75%{
  transform:scale(.95) translateX(-50%);

}
100%{
  transform:scale(1) translateX(-50%);

}
`

export const StyledBadge = styled.span `

    line-height: 1.5;
    box-sizing: border-box;
    margin: 0;
    padding: 0;
    list-style: none;
    position: relative;
    display: inline-flex;
    flex-direction: column;
    line-height: 1;
    vertical-align: middle;
    z-index: ${({theme}) => theme.zIndex.badge};



`

export const StyledSup = styled.sup `
  background: ${props => props.backgroundColor
  ? props.backgroundColor
  : props.theme.primary.main};
  position: absolute;
  z-index: ${({theme}) => theme.zIndex.badge};

  transform: scale(1) translateX(-50%);
  top: -1rem;
  left: ${props => props.widthChildren
    ? `${props.widthChildren / 10}rem`
    : 0}
  height: 2rem;
  border-radius: 1rem;
  min-width: 2rem;
  color: #fff;
  line-height: 2rem;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-wrap:wrap;
  padding: ${props => props.count < 10
      ? '0 .6rem'
      : '0 .8rem'};
  font-size: 1.2rem;
  font-weight: 400;
  white-space: nowrap;
  margin-top: ${props => `${props.offset[0] / 10}rem`};
  margin-left:  ${props => `${props.offset[1] / 10}rem`};
  animation: ${bouncing} .3s 1 linear both;



  transform-origin: -10% center;


`
