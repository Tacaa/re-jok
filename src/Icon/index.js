// @flow
import * as React from 'react';

import FontAwesome from 'react-fontawesome';
import styled from 'styled-components'

const IconContainer = styled.div`
order: ${props => props.iconPosition === "left" ? 0 : 1};
color:  ${props => props.color ? props.theme[props.color].main : null};
display: inline-block;


`

type Props = {
  /**name of icon**/
  name: string,
  /** Size of icon shuold be string like lg, 2x, 3x v..v if you don't want use size you can custom with style**/
  size?: string,
  /**Override style of icon**/
  style?: Object,
  /**Add more class to icon**/
  className?: string,
  /** Set icon spin or not**/
  spin: boolean,
  /** Set color to icon but it just set color like error, warning,success.. i recomment you set in style or class**/
  color?: string,
  /** set postion of icon right or left**/
  iconPosition: 'left' | 'right',
}

const defaultProps = {
  spin: false,
  iconPosition: "left"
}


class Icon extends React.Component<Props> {
  static defaultProps = defaultProps;

  render(){
    const {
      iconPosition,
      name,
      size,
      style,
      className,
      spin,
      color,
    } = this.props;
    return (
      <IconContainer color={color} iconPosition={iconPosition}>
        <FontAwesome
         name={name}
         size={size}
         style={style}
         className={className}
         spin={spin}

        />
      </IconContainer>
    )
  }
}



export default Icon;
