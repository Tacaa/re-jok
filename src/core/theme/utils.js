import {
  defaultTheme
} from './index';


export const configTheme = (theme: Object) => {
  let jokTheme = {...defaultTheme, ...theme};
  return jokTheme;
}
