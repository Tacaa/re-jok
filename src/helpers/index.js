export const generalId = () => {
  return '_' + Math.random().toString(36).substr(2, 9);
}

//Default Function
export const noop = () => {}
