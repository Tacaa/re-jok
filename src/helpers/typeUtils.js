/** Check element has child or not**/
export const isChild = children => children === null
    || children === undefined
    || (Array.isArray(children) && children.length === 0)

/** Check undefined**/

export const isUndef = value => typeof value === 'undefined';
/** Check function or not**/

export const isFunction = value => !isUndef(value) && typeof value === 'function';

/** Check Object or not**/
export const isObject = value =>
      typeof value === 'object' && value !== null && !Array.isArray(value);

/** Check Array or not**/
export const isArray = value => Array.isArray(value);

/** Check Number or not**/
export const isNumber = value => Number.isInteger(value) && typeof value === "number"
