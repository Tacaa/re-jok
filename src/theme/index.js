// @flow
// export our theme to the provider
// This is globally available in styled-components when interpolating a function like so:
// ${(props) => props.theme}
// Or using import { withTheme } from 'styled-components';


export const defaultTheme = {
  size: {
    small: {
      fontSize: '1.2rem',
      padding: '0.5rem 0.9rem',
      "lineHeight": "1",
      "borderRadius": ".2rem",
      width: "2.4rem",
      height: "2.4rem"
    },
    default: {
      fontSize: '1.4rem',
      padding: '0.9rem 1.8rem',
      lineHeight: "1.5",
      borderRadius: ".4rem",
      width: "3.2rem",
      height: "3.2rem"
    },
    medium: {
      fontSize: '1.8rem',
      padding: '1.4rem 2.7rem',
      lineHeight: "1.5",
      borderRadius: ".6rem",
      width: "4rem",
      height: "4rem"

    },
    large: {
      fontSize: '2rem',
      padding: '1.8rem 3.6rem',
      lineHeight: "2",
      borderRadius: ".6rem",
      width: "4.8rem",
      height: "4.8rem"
    }
  },
  grid:{
    xs: 768,
    sm: 768,
    md: 992,
    lg: 1200,
    xl: 1920
  },
  text: {
    primary: "rgba(0, 0, 0, 0.85)",
    secondary: "rgba(0, 0, 0, 0.48)",
    hint: "rgba(0, 0, 0, 0.38)"
  },
  background: {
    paper: "#fff",
    default: "#fafafa"
  },
  typography: {
    title: {
      fontSize: "2rem",
      fontWeight: "600",
      lineHeight: "2.8rem"

    },
    subtitle: {
      fontSize: "1.6rem",
      fontWeight: "400",
      lineHeight: "2.4rem"
    }
  },
  primary: {
    main: "#1c8ffb",

  },
  secondary: {
    main: "#1b1c1d",
  },
  info: {
    main: "#1890ff"
  },
  error: {
    main: "#f5222d"
  },
  success: {
    main: "#52c41a"
  },
  warning: {
    main: "#faad14"
  },
  zIndex: {
    base: 1,
    background: 0,
    hidden: -1,
    card: 2,
    loading: 3,
    avatar: 3,
    form: 3,
    search: 3,
    badge: 100,
    modal: 1500,
  }
};

const changeTheme = (defaultTheme, theme)=>{
  let cloneTheme = {...defaultTheme};


  Object.keys(theme).forEach(function (key) {
    if ( 'object' === typeof theme[key] && !Array.isArray(theme[key]) ) {
    	changeTheme(cloneTheme[key], theme[key])
    }
    else if ( 'string' === typeof theme[key] || 'number' === typeof theme[key] ) {
      cloneTheme[key] = theme[key]
    }
  })
  return cloneTheme;
}

export const setTheme = (theme: Object) => {

   const newTheme = changeTheme(defaultTheme, theme);

   return newTheme;

}
