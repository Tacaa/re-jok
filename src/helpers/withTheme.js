import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import {defaultTheme} from '../theme';
import { ThemeProvider } from 'styled-components';
export function renderWithTheme(component) {

  return renderer.create(
    <ThemeProvider theme={defaultTheme}>
      {component}
    </ThemeProvider>
  );
};

export const shallowWithTheme = (tree, theme) => {
  const context = shallow(<ThemeProvider theme={theme} />)
    .instance()
    .getChildContext()
  return shallow(tree, { context })
};

export const mountWithTheme = (tree, theme) => {
  const context = shallow(<ThemeProvider theme={theme} />)
    .instance()
    .getChildContext()

  return mount(tree, {
    context,
    childContextTypes: ThemeProvider.childContextTypes
  })
};
