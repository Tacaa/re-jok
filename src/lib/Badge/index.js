// @flow
import * as React from 'react';

import ReactDOM from 'react-dom';

import {StyledBadge, StyledSup} from './style';

import { isChild } from '../helpers/typeUtils';


type Props = {
  /** number will display in badge**/
  count: number,
  /** children should be html element and just one child**/
  children: React$Element<any>,
  /** Max number to show**/
  maxCount: number,
  /** set position of badge**/
  offset: Array<number>,
  /** Set background color of badge**/
  backgroundColor?: string
}

type State = {
  /** Width of children element**/
  widthChildren: number
}

const defaultProps = {
  count: 0,
  maxCount: 99,
  offset: [0, 0]
}

class Badge extends React.Component<Props, State> {
  static defaultProps = defaultProps;

  constructor(props : Props) {
    super(props);
    this.state = {
      widthChildren: 0
    }

  }

  refChildren: {
    current: null | React$ElementRef<any>
  } = React.createRef();

  componentDidMount() {
    const node = ReactDOM.findDOMNode(this.refChildren.current);

    if (node && node instanceof HTMLElement) {
      this.setState({widthChildren: node.offsetWidth})
    }

  }

  render() {
    const {count, children, maxCount} = this.props;

    const {widthChildren} = this.state;

    const hasChild = !isChild(children);

    const label = count > maxCount
      ? `${maxCount}+`
      : count;

    return (<StyledBadge>
      {
        hasChild && React.Children.count(children) === 1
          ? <children.type {...children.props} ref={this.refChildren}/>
          : null
      }
      <StyledSup widthChildren={widthChildren} {...this.props}>{label}</StyledSup>
    </StyledBadge>)
  }
}

export default Badge
