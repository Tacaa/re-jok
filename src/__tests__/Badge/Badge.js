import React from 'react';
import Badge from '../../Badge'
import {defaultTheme} from '../../theme'
import Icon from '../../Icon';
import {shallowWithTheme, mountWithTheme, renderWithTheme} from '../../helpers/withTheme';
import 'jest-enzyme'
import 'jest-styled-components'

describe('<Badge />', () => {
  it('should render Badge', () => {
    const tree = renderWithTheme(<Badge></Badge>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render Badge correctly with default props', () => {
    const wrapper = mountWithTheme(<Badge/>, defaultTheme);
    expect(wrapper.props().count).toEqual(0);
    expect(wrapper.props().maxCount).toEqual(99);
    expect(wrapper.props().offset).toEqual([0, 0]);

  });

})
