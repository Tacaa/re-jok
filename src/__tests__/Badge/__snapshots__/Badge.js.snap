// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Badge /> should render Badge 1`] = `
.c0 {
  line-height: 1.5;
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  list-style: none;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  line-height: 1;
  vertical-align: middle;
  z-index: 100;
}

.c1 {
  background: #1c8ffb;
  position: absolute;
  z-index: 100;
  -webkit-transform: scale(1) translateX(-50%);
  -ms-transform: scale(1) translateX(-50%);
  transform: scale(1) translateX(-50%);
  top: -1rem;
  left: 0;
  height: 2rem;
  border-radius: 1rem;
  min-width: 2rem;
  color: #fff;
  line-height: 2rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  padding: 0 .6rem;
  font-size: 1.2rem;
  font-weight: 400;
  white-space: nowrap;
  margin-top: 0rem;
  margin-left: 0rem;
  -webkit-animation: jlLOYG .3s 1 linear both;
  animation: jlLOYG .3s 1 linear both;
  -webkit-transform-origin: -10% center;
  -ms-transform-origin: -10% center;
  transform-origin: -10% center;
}

<span
  className="c0"
>
  <sup
    className="c1"
    offset={
      Array [
        0,
        0,
      ]
    }
  >
    0
  </sup>
</span>
`;
